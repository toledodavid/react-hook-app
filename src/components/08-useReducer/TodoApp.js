import React, {useReducer} from 'react';
import { todoReducer } from './todoReducer';
import './styles.css';
import { useForm } from '../../hooks/useForm';


const initialState = [
  {
    id: new Date().getTime(),
    desc: 'Learn React',
    done: false
  }
];


const TodoApp = () => {

  const [ todos, dispatch ] = useReducer(todoReducer, initialState);

  const [{description}, handleInputChange, reset] = useForm({
    description: ''
  });

  console.log(description);

  const handleSubmit = (event) => {
    event.preventDefault();

    if (description.trim().length <= 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false
    }

    const action = {
      type: 'add',
      payload: newTodo
    }

    dispatch(action);
    reset();
  }

  return(
    <div>
      <h1>Todo App ({todos.length})</h1>
      <hr/>

      <div className="row">

        <div className="col-7">

          <ul className="list-group list-group-flush">
            {
              todos.map((todo, index) => (
                <li key={todo.id}
                    className="list-group-item">
                  <p className="text-center">{index + 1}. {todo.desc}</p>
                  <button className="btn btn-danger">Delete</button>
                </li>
              ))
            }
          </ul>

        </div>

        <div className="col-5">
          <h4>Add TODO</h4>
          <hr/>

          <form onSubmit={handleSubmit}>
            <input type="text" 
                    name="description"
                    placeholder="Learn..."
                    autoComplete="off"
                    className="form-control"
                    value={description}
                    onChange={handleInputChange} />

            <button type="submit" className="btn btn-outline-primary btn-block mt-1">Add</button>
          </form>
        </div>

      </div>
    </div>
  );
}



export default TodoApp;