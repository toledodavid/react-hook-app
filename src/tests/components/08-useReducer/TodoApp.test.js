import '@testing-library/jest-dom';
import { act } from '@testing-library/react';
import { mount, shallow } from 'enzyme';
import TodoApp from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from '../../fixtures/demoTodos';


describe('Tests in <TodoApp /> component', () => {

  const wrapper = shallow(<TodoApp/>);

  Storage.prototype.setItem = jest.fn();

  test('It should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('It should add two new TODOs', () => {
    const wrapper = mount(<TodoApp />);

    act(() => {
      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1]);
    });

    expect(wrapper.find('h1').text().trim()).toBe('Todo App (2)');
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  });

  test('It should delete a TODO', () => {
    wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
    wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);

    expect(wrapper.find('h1').text().trim()).toBe('Todo App (0)');
  });

});