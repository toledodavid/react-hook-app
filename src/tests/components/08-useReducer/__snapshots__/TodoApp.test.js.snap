// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <TodoApp /> component It should render correctly 1`] = `
<div>
  <h1>
    Todo App (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
